File: sdk/runanywhere-python/runanywhere/events.py
"""Event types for every stream, plus the SDK-level event bus.

One grammar everywhere: a stream emits ``started``, then deltas, then ``completed`` — or
raises into the consumer. Each event family is a single dataclass carrying a ``kind`` so a
consumer can switch on it without isinstance chains.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Callable, Dict, List, Optional

from .inputs import ModelCategory
from .results import (
    GenerationResult,
    ImageResult,
    Match,
    ModelInfo,
    RagResult,
    TokenKind,
    ToolCall,
    Transcription,
)

__all__ = [
    "DownloadEvent",
    "DownloadEventKind",
    "EventBus",
    "GenerationEvent",
    "GenerationEventKind",
    "ImageEvent",
    "ImageEventKind",
    "RagEvent",
    "RagEventKind",
    "SdkEvent",
    "SdkEventKind",
    "TranscriptionEvent",
    "TranscriptionEventKind",
    "VadEvent",
    "VadEventKind",
    "VoiceAgentState",
    "VoiceEvent",
    "VoiceEventKind",
    "bus",
]


class GenerationEventKind(IntEnum):
    """Stages of a generation stream."""

    STARTED = 0
    TOKEN = 1
    TOOL_CALL = 2
    COMPLETED = 3


@dataclass
class GenerationEvent:
    """One event of an ``llm``/``vlm`` stream."""

    kind: GenerationEventKind
    request_id: str = ""
    text: str = ""
    token_kind: TokenKind = TokenKind.TEXT
    tool_call: Optional[ToolCall] = None
    result: Optional[GenerationResult] = None

    @property
    def is_token(self) -> bool:
        """True for a token delta."""
        return self.kind == GenerationEventKind.TOKEN

    @property
    def is_completed(self) -> bool:
        """True for the terminal event, whose ``result`` carries the metrics."""
        return self.kind == GenerationEventKind.COMPLETED

    @property
    def is_thought(self) -> bool:
        """True when this token is the model's thinking rather than its answer."""
        return self.kind == GenerationEventKind.TOKEN and self.token_kind == TokenKind.THOUGHT


class TranscriptionEventKind(IntEnum):
    """Stages of a transcription stream."""

    STARTED = 0
    PARTIAL = 1
    FINAL = 2


@dataclass
class TranscriptionEvent:
    """One event of an ``stt`` stream."""

    kind: TranscriptionEventKind
    text: str = ""
    result: Optional[Transcription] = None


class VadEventKind(IntEnum):
    """Stages of a speech-detection stream."""

    SPEECH_STARTED = 0
    SPEECH_ENDED = 1


@dataclass
class VadEvent:
    """One speech boundary detected in an audio stream."""

    kind: VadEventKind
    timestamp_ms: int = 0


class VoiceAgentState(IntEnum):
    """What a voice session's agent is doing."""

    LISTENING = 0
    THINKING = 1
    SPEAKING = 2


class VoiceEventKind(IntEnum):
    """Kinds of voice-session event."""

    USER_TRANSCRIBED = 0
    AGENT_STATE_CHANGED = 1
    AGENT_RESPONSE = 2
    SPEECH_STARTED = 3
    SPEECH_ENDED = 4
    ERROR = 5


@dataclass
class VoiceEvent:
    """One event emitted by a voice session."""

    kind: VoiceEventKind
    text: str = ""
    is_final: bool = False
    state: Optional[VoiceAgentState] = None
    message: str = ""
    recoverable: bool = True


class RagEventKind(IntEnum):
    """Stages of a RAG query stream."""

    RETRIEVED = 0
    TOKEN = 1
    COMPLETED = 2


@dataclass
class RagEvent:
    """One event of a ``RagSession.query_stream``."""

    kind: RagEventKind
    matches: List[Match] = field(default_factory=list)
    text: str = ""
    token_kind: TokenKind = TokenKind.TEXT
    result: Optional[RagResult] = None

    @property
    def is_token(self) -> bool:
        """True for an answer-token delta."""
        return self.kind == RagEventKind.TOKEN

    @property
    def is_completed(self) -> bool:
        """True for the terminal event, whose ``result`` carries the answer."""
        return self.kind == RagEventKind.COMPLETED


class ImageEventKind(IntEnum):
    """Stages of an image generation stream."""

    STARTED = 0
    PROGRESS = 1
    COMPLETED = 2


@dataclass
class ImageEvent:
    """One event of an ``images`` stream."""

    kind: ImageEventKind
    step: int = 0
    total_steps: int = 0
    partial_image: Optional[bytes] = None
    result: Optional[ImageResult] = None


class DownloadEventKind(IntEnum):
    """Stages of a model download."""

    PROGRESS = 0
    EXTRACTING = 1
    COMPLETED = 2


@dataclass
class DownloadEvent:
    """One event of ``models.download``."""

    kind: DownloadEventKind
    file: str = ""
    bytes_done: int = 0
    bytes_total: int = 0
    percent: int = 0
    model: Optional[ModelInfo] = None


class SdkEventKind(IntEnum):
    """Kinds of SDK-level breadcrumb."""

    READY = 0
    MODEL_LOADED = 1
    MODEL_UNLOADED = 2
    ERROR = 3


@dataclass(frozen=True)
class SdkEvent:
    """A lifecycle, model or error breadcrumb published on :data:`bus`."""

    kind: SdkEventKind
    model_id: Optional[str] = None
    category: Optional[ModelCategory] = None
    message: str = ""
    recoverable: bool = True


EventListener = Callable[[SdkEvent], None]


class EventBus:
    """Pub/sub bus for :class:`SdkEvent` where a throwing listener never breaks an emit."""

    def __init__(self) -> None:
        # dict preserves insertion order and gives set-like membership; the value is unused.
        self._listeners: Dict[EventListener, None] = {}

    def on(self, listener: EventListener) -> Callable[[], None]:
        """Subscribe to all events; returns the unsubscribe function."""
        self._listeners[listener] = None

        def off() -> None:
            self._listeners.pop(listener, None)

        return off

    def once(self, listener: EventListener) -> Callable[[], None]:
        """Subscribe to the next event only; returns the unsubscribe function."""

        def wrapper(event: SdkEvent) -> None:
            off()
            listener(event)

        off = self.on(wrapper)
        return off

    def off(self, listener: EventListener) -> None:
        """Unsubscribe a listener (no-op if it was never registered)."""
        self._listeners.pop(listener, None)

    def emit(self, event: SdkEvent) -> None:
        """Publish an event to every listener."""
        for listener in list(self._listeners):
            try:
                listener(event)
            except Exception:  # noqa: BLE001
                # A misbehaving listener must not disrupt the others. Catch Exception (NOT
                # BaseException) so KeyboardInterrupt / SystemExit / CancelledError still
                # propagate. Log the traceback only — never the event payload.
                logging.getLogger(__name__).exception("EventBus listener failed")

    def remove_all(self) -> None:
        """Drop every registered listener."""
        self._listeners.clear()

    @property
    def listener_count(self) -> int:
        """Number of currently registered listeners."""
        return len(self._listeners)


#: The process-wide bus exposed as ``runanywhere.events``.
bus = EventBus()

File: bindings/flutter/packages/runanywhere/lib/public/api/types/events.dart
// SPDX-License-Identifier: Apache-2.0
//
// Public v4 event grammar: one shape throughout — `started`, then deltas,
// then a terminal `completed`/`failed`/`cancelled`. A stream never fabricates
// a successful `completed` after an in-flight failure or cancellation; the
// terminal event carries whatever partial output was produced instead.
//
// Reference: bindings/swift/Sources/RunAnywhere/Public/API/Events.swift,
// bindings/web/packages/core/src/Public/API/Events.ts, and
// .claude/handoff/plans/public_api_spec.md.

import 'dart:typed_data';

import 'package:runanywhere/foundation/errors/sdk_exception.dart';
import 'package:runanywhere/generated/model_types.pb.dart' show ModelInfo;
import 'package:runanywhere/generated/model_types.pbenum.dart'
    show ModelCategory;
import 'package:runanywhere/generated/tool_calling.pb.dart' show ToolCall;
import 'package:runanywhere/public/api/types/results.dart';

/// Whether a streamed delta is answer text or chain-of-thought.
enum TokenKind {
  /// Answer content the user should see.
  text,

  /// Chain-of-thought content, emitted only when reasoning output is enabled.
  thought,
}

/// Events emitted by `llm.generateStream` and `vlm.generateStream`.
sealed class GenerationEvent {
  const GenerationEvent();
}

/// The request was admitted and generation began.
final class GenerationStarted extends GenerationEvent {
  /// Announce a request id for correlation.
  const GenerationStarted(this.requestId);

  /// Correlation id for the generation.
  final String requestId;
}

/// One answer-text delta.
final class GenerationTextDelta extends GenerationEvent {
  /// Carry the text produced since the previous delta.
  const GenerationTextDelta(this.text, {this.requestId = ''});

  /// Text produced since the previous delta.
  final String text;

  /// Correlation id for the generation.
  final String requestId;
}

/// One reasoning/thinking delta, emitted only when reasoning output is
/// enabled.
final class GenerationReasoningDelta extends GenerationEvent {
  /// Carry the reasoning text produced since the previous delta.
  const GenerationReasoningDelta(this.text, {this.requestId = ''});

  /// Reasoning text produced since the previous delta.
  final String text;

  /// Correlation id for the generation.
  final String requestId;
}

/// Deprecated v3 shape: one decoded token tagged with [kind] instead of a
/// dedicated event per stream.
///
/// `llm`/`vlm` `generateStream` no longer emit this — they emit
/// [GenerationTextDelta] (`kind == TokenKind.text`) or
/// [GenerationReasoningDelta] (`kind == TokenKind.thought`) — but the type
/// remains for source compatibility.
@Deprecated(
  'Use GenerationTextDelta for answer text or GenerationReasoningDelta '
  'for thoughts',
)
final class GenerationToken extends GenerationEvent {
  /// Carry a token and its semantic category.
  const GenerationToken(
    this.text, {
    this.kind = TokenKind.text,
    this.requestId = '',
  });

  /// Token text.
  final String text;

  /// Whether this is answer text or a thought.
  final TokenKind kind;

  /// Correlation id for the generation.
  final String requestId;
}

/// The model asked to call a tool.
final class GenerationToolCallAdded extends GenerationEvent {
  /// Carry the requested call.
  const GenerationToolCallAdded(this.call, {this.requestId = ''});

  /// Tool name and arguments the model produced.
  final ToolCall call;

  /// Correlation id for the generation.
  final String requestId;
}

/// Deprecated v3 name for [GenerationToolCallAdded].
@Deprecated('Use GenerationToolCallAdded')
final class GenerationToolCall extends GenerationEvent {
  /// Carry the requested call.
  const GenerationToolCall(this.call, {this.requestId = ''});

  /// Tool name and arguments the model produced.
  final ToolCall call;

  /// Correlation id for the generation.
  final String requestId;
}

/// Token accounting for the request so far.
final class GenerationUsage extends GenerationEvent {
  /// Carry running token counts.
  const GenerationUsage({
    required this.inputTokens,
    required this.outputTokens,
    this.requestId = '',
  });

  /// Prompt tokens consumed.
  final int inputTokens;

  /// Completion tokens produced so far.
  final int outputTokens;

  /// Correlation id for the generation.
  final String requestId;
}

/// Decoding finished; carries the aggregate result and metrics. Terminal —
/// no event follows.
final class GenerationCompleted extends GenerationEvent {
  /// Carry the terminal result. [requestId] defaults to `result.requestId`
  /// when not given explicitly.
  GenerationCompleted(this.result, {String? requestId})
    : requestId = requestId ?? result.requestId;

  /// Full text, tool calls, and metrics.
  final GenerationResult result;

  /// Correlation id for the generation.
  final String requestId;
}

/// Generation failed in flight. Terminal — no event follows.
///
/// `llm`/`vlm` `generateStream` never fabricate a [GenerationCompleted]
/// after this.
final class GenerationFailed extends GenerationEvent {
  /// Carry the failure and any partial text produced before it.
  const GenerationFailed(this.error, {this.partial, this.requestId = ''});

  /// What went wrong.
  final SDKException error;

  /// Text produced before the failure, when any was.
  final String? partial;

  /// Correlation id for the generation.
  final String requestId;
}

/// Generation was cancelled by the caller. Terminal — no event follows.
final class GenerationCancelled extends GenerationEvent {
  /// Carry any partial text produced before cancellation.
  const GenerationCancelled({this.partial, this.requestId = ''});

  /// Text produced before cancellation, when any was.
  final String? partial;

  /// Correlation id for the generation.
  final String requestId;
}

/// Events emitted by `stt.transcribeStream`.
sealed class TranscriptionEvent {
  const TranscriptionEvent();
}

/// The session opened and is waiting for audio.
final class TranscriptionStarted extends TranscriptionEvent {
  /// Announce session start, optionally with a correlation id.
  const TranscriptionStarted([this.requestId = '']);

  /// Correlation id for this stream, when one was assigned.
  final String requestId;
}

/// An in-progress hypothesis that may still change.
final class TranscriptionPartial extends TranscriptionEvent {
  /// Carry the current hypothesis.
  const TranscriptionPartial(this.text);

  /// Hypothesis text so far.
  final String text;
}

/// The session closed with a settled transcript.
final class TranscriptionFinal extends TranscriptionEvent {
  /// Carry the settled transcript.
  const TranscriptionFinal(this.transcription);

  /// Final transcript with timings.
  final Transcription transcription;
}

/// The stream settled — [TranscriptionFinal] (when the session produced
/// one) has already been emitted. Terminal.
final class TranscriptionCompleted extends TranscriptionEvent {
  /// Announce a clean finish.
  const TranscriptionCompleted();
}

/// The stream ended because of an error. Terminal — no event follows.
///
/// `stt.openStream` never fabricates a [TranscriptionFinal]/completion when
/// the native session ends without one; this is what it emits instead.
final class TranscriptionFailed extends TranscriptionEvent {
  /// Carry the failure.
  const TranscriptionFailed(this.error);

  /// What went wrong.
  final SDKException error;
}

/// The stream was cancelled before it produced a final transcript. Terminal.
final class TranscriptionCancelled extends TranscriptionEvent {
  /// Announce cancellation.
  const TranscriptionCancelled();
}

/// Events emitted by `vad.openStream`.
sealed class VadEvent {
  const VadEvent();
}

/// Speech began after a run of silence (or stream start).
final class VadSpeechStarted extends VadEvent {
  /// Carry the frame timestamp that crossed the boundary, when known.
  const VadSpeechStarted(this.timestampMs);

  /// Timestamp of the triggering frame.
  final int? timestampMs;
}

/// Speech ended after a run of activity.
final class VadSpeechEnded extends VadEvent {
  /// Carry the frame timestamp that crossed the boundary, when known.
  const VadSpeechEnded(this.timestampMs);

  /// Timestamp of the triggering frame.
  final int? timestampMs;
}

/// One frame's verdict, with no state transition implied.
final class VadActivity extends VadEvent {
  /// Carry one frame's detection result.
  const VadActivity({
    required this.isSpeech,
    required this.probability,
    this.timestampMs,
  });

  /// True when this frame was classified as speech.
  final bool isSpeech;

  /// Speech probability for this frame.
  final double probability;

  /// Timestamp of this frame.
  final int? timestampMs;
}

/// The stream failed in flight. Terminal — no event follows.
final class VadFailed extends VadEvent {
  /// Carry the failure.
  const VadFailed(this.error);

  /// What went wrong.
  final SDKException error;
}

/// The stream finished normally, after `finish()` drained the last frame.
final class VadCompleted extends VadEvent {
  /// Announce a clean finish.
  const VadCompleted();
}

/// What the voice agent is doing right now.
enum AgentState {
  /// Waiting for the user to speak.
  listening,

  /// Generating a reply.
  thinking,

  /// Speaking the reply.
  speaking,
}

/// Events emitted by a [VoiceEvent] stream on a voice session.
sealed class VoiceEvent {
  const VoiceEvent();
}

/// The user's speech was transcribed.
final class VoiceUserTranscribed extends VoiceEvent {
  /// Carry a user hypothesis or settled transcript.
  const VoiceUserTranscribed(this.text, {required this.isFinal});

  /// Transcribed text.
  final String text;

  /// True when the transcript will not change again.
  final bool isFinal;
}

/// The agent moved between listening, thinking, and speaking.
final class VoiceAgentStateChanged extends VoiceEvent {
  /// Carry the new state.
  const VoiceAgentStateChanged(this.state);

  /// What the agent is doing now.
  final AgentState state;
}

/// The agent produced reply text.
final class VoiceAgentResponse extends VoiceEvent {
  /// Carry the reply text.
  const VoiceAgentResponse(this.text);

  /// Reply text, accumulated so far.
  final String text;
}

/// Speech was detected on the microphone.
final class VoiceSpeechStarted extends VoiceEvent {
  /// Announce speech onset.
  const VoiceSpeechStarted();
}

/// Speech on the microphone ended.
final class VoiceSpeechEnded extends VoiceEvent {
  /// Announce speech offset.
  const VoiceSpeechEnded();
}

/// A component of the session hit trouble.
final class VoiceError extends VoiceEvent {
  /// Carry the failure and whether the session survives it.
  const VoiceError(this.message, {required this.recoverable});

  /// What went wrong.
  final String message;

  /// True when the session keeps running.
  final bool recoverable;
}

/// Events emitted by `RagSession.queryStream`.
sealed class RagEvent {
  const RagEvent();
}

/// Retrieval finished and produced these chunks.
final class RagRetrieved extends RagEvent {
  /// Carry the retrieved chunks.
  const RagRetrieved(this.matches);

  /// Chunks that will ground the answer.
  final List<Match> matches;
}

/// One decoded answer token.
final class RagToken extends RagEvent {
  /// Carry a token and its semantic category.
  const RagToken(this.text, {this.kind = TokenKind.text});

  /// Token text.
  final String text;

  /// Whether this is answer text or a thought.
  final TokenKind kind;
}

/// The query finished; carries the grounded answer.
final class RagCompleted extends RagEvent {
  /// Carry the terminal result.
  const RagCompleted(this.result);

  /// Answer, sources, and metrics.
  final RagResult result;
}

/// Events emitted by `images.generateStream`.
sealed class ImageEvent {
  const ImageEvent();
}

/// Sampling began.
final class ImageStarted extends ImageEvent {
  /// Announce the start of sampling.
  const ImageStarted();
}

/// One denoising step completed.
final class ImageProgress extends ImageEvent {
  /// Carry step counters and an optional preview.
  const ImageProgress({
    required this.step,
    required this.totalSteps,
    this.partialImage,
  });

  /// Steps completed.
  final int step;

  /// Steps in the schedule.
  final int totalSteps;

  /// Preview image, when partial reporting is enabled.
  final Uint8List? partialImage;
}

/// Sampling finished; carries the images.
final class ImageCompleted extends ImageEvent {
  /// Carry the terminal result.
  const ImageCompleted(this.result);

  /// Generated images and the sampling parameters used.
  final ImageResult result;
}

/// Events emitted by `models.download`, correlated by [DownloadEvent.operationId]
/// and its per-event `sequence` counter.
///
/// One grammar throughout: `started`, then progress deltas, then a terminal
/// `completed`/`failed`/`cancelled`. `models.download` never fabricates a
/// [DownloadCompleted] after a [DownloadFailed]/[DownloadCancelled].
sealed class DownloadEvent {
  const DownloadEvent();
}

/// The transfer was admitted and began.
final class DownloadStarted extends DownloadEvent {
  /// Announce the start of one download operation.
  const DownloadStarted(this.operationId, {this.sequence = 0});

  /// Correlation id for this download, stable across every event it emits.
  final String operationId;

  /// Monotonic per-operation event counter.
  final int sequence;
}

/// Bytes are arriving.
final class DownloadProgressEvent extends DownloadEvent {
  /// Carry transfer counters and commons-owned overall progress.
  const DownloadProgressEvent({
    required this.operationId,
    required this.bytesDone,
    required this.bytesTotal,
    this.sequence = 0,
    this.file,
    this.overallProgress,
  });

  /// Correlation id for this download.
  final String operationId;

  /// Bytes received so far.
  final int bytesDone;

  /// Total bytes expected. Zero when the server did not report a length.
  final int bytesTotal;

  /// Monotonic per-operation event counter.
  final int sequence;

  /// File currently being fetched, for multi-file downloads.
  final String? file;

  /// Fraction of the whole download that is done, `0.0..1.0`, as reported by
  /// commons (`DownloadProgress.overall_progress`). Null when commons has not
  /// yet reported a positive value — never re-derived from [bytesDone]/
  /// [bytesTotal] (multi-file plans make that a per-file figure).
  final double? overallProgress;
}

/// Downloaded bytes are being checksummed/validated.
final class DownloadVerifying extends DownloadEvent {
  /// Announce the verification phase.
  const DownloadVerifying(this.operationId, {this.sequence = 0});

  /// Correlation id for this download.
  final String operationId;

  /// Monotonic per-operation event counter.
  final int sequence;
}

/// The transfer finished and the archive is being unpacked.
final class DownloadExtracting extends DownloadEvent {
  /// Announce the extraction phase.
  const DownloadExtracting(this.operationId, {this.sequence = 0, this.percent});

  /// Correlation id for this download.
  final String operationId;

  /// Monotonic per-operation event counter.
  final int sequence;

  /// Extraction completion, when the backend reports incremental progress.
  final double? percent;
}

/// The model is on disk and registered. Terminal — no event follows.
final class DownloadCompleted extends DownloadEvent {
  /// Carry the registered model.
  const DownloadCompleted(this.operationId, this.model, {this.sequence = 0});

  /// Correlation id for this download.
  final String operationId;

  /// Registry entry with its resolved local path.
  final ModelInfo model;

  /// Monotonic per-operation event counter.
  final int sequence;
}

/// The transfer failed in flight. Terminal — no event follows.
final class DownloadFailed extends DownloadEvent {
  /// Carry the failure.
  const DownloadFailed(this.operationId, this.error, {this.sequence = 0});

  /// Correlation id for this download.
  final String operationId;

  /// What went wrong.
  final SDKException error;

  /// Monotonic per-operation event counter.
  final int sequence;
}

/// The transfer was cancelled by the caller. Terminal — no event follows.
final class DownloadCancelled extends DownloadEvent {
  /// Announce cancellation.
  const DownloadCancelled(this.operationId, {this.sequence = 0});

  /// Correlation id for this download.
  final String operationId;

  /// Monotonic per-operation event counter.
  final int sequence;
}

/// Lifecycle, download, and error breadcrumbs from `RunAnywhere.events`.
sealed class SdkEvent {
  const SdkEvent();
}

/// Local inference became usable.
final class SdkReady extends SdkEvent {
  /// Announce readiness.
  const SdkReady();
}

/// A model finished loading.
final class SdkModelLoaded extends SdkEvent {
  /// Carry the loaded model's identity.
  const SdkModelLoaded(this.id, this.category);

  /// Model id.
  final String id;

  /// Category the model was loaded under.
  final ModelCategory category;
}

/// A model was unloaded.
final class SdkModelUnloaded extends SdkEvent {
  /// Carry the unloaded model's id.
  const SdkModelUnloaded(this.id);

  /// Model id.
  final String id;
}

/// Something went wrong outside a caller's request.
final class SdkError extends SdkEvent {
  /// Carry the failure and whether the SDK survives it.
  const SdkError(this.message, {required this.recoverable});

  /// What went wrong.
  final String message;

  /// True when the SDK keeps working.
  final bool recoverable;
}
